tests/api_client/test_api_client.py

```diff
@@ -0,0 +1,309 @@
+                "later.")
+        },
+    }
+
+    status_code = None
+
+    def __init__(self, status_code):
+        self.status_code = status_code
+
+    def json(self):
+        return self.possible_responses[self.status_code]
+
+    def raise_for_status(self):
+        raise requests.exceptions.HTTPError()
+
+
+@pytest.mark.api_client
+class TestAPIClient:
+    def test_init_default_client(self):
+        client = api_client.APIClient()
+        assert client.USE_SSL is True
+        assert client.AUTHENTICATION_TOKEN is None
+        assert client.BASE_URL == 'https://localhost'
+        assert client.HEADERS == {}
+
+    def test_init_default_client_no_ssl(self):
+        client = api_client.APIClient(use_ssl=False)
+        assert client.USE_SSL is False
+        assert client.AUTHENTICATION_TOKEN is None
+        assert client.BASE_URL == 'http://localhost'
+        assert client.HEADERS == {}
+
+    def test_init_custom_token_client(self):
+        test_token = 'TEST'
+        client = api_client.APIClient(authentication_token=test_token)
+        assert client.AUTHENTICATION_TOKEN == test_token
+
+    def test_load_configuration(self, client):
+        with pytest.raises(NotImplementedError):
+            client.load_configuration_from_file()
+
+    def test_authenticate(self, client):
+        with pytest.raises(NotImplementedError):
+            username = 'TEST_USER'
+            password = 'TEST_PASSWORD'
+            client.authenticate(username, password)
+
+    def test_join_path(self, client):
+        assert client.join_path('jobs') == f'{client.BASE_URL}/jobs'
+
+
+@pytest.mark.api_client
+class TestResourceManager:
+    def test_init(self):
+        resource = MagicMock()
+        client = MagicMock()
+        manager = ResourceManager(resource, client)
+
+        assert manager.resource == resource
+        assert manager.client == client
+
+    def test_join_path(self):
+        mock_resource = MagicMock()
+        mock_resource.PATH = 'some-path'
+
+        manager = ResourceManager(mock_resource, MagicMock())
+        assert manager.join_path('test') == "some-path/test"
+
+    def test_get_unsupported(self):
+        mock_resource = MagicMock()
+        mock_resource.SUPPORTED_METHODS = ()
+        manager = ResourceManager(mock_resource, MagicMock())
+        with pytest.raises(MethodNotSupportedException):
+            manager.get(1)
+
+    def test_get(self, monkeypatch):
+        mock_resource = MagicMock()
+        mock_client = MagicMock()
+        mock_response = MagicMock()
+        mock_client.get = MagicMock(return_value=mock_response)
+
+        mock_resource.SUPPORTED_METHODS = ('GET',)
+
+        manager = ResourceManager(mock_resource, mock_client)
+        monkeypatch.setattr(manager, "handle_response", MagicMock())
+
+        manager.get(1)
+
+        # TODO test that this is called with correct path
+        mock_client.get.assert_called_once()
```

Comment:
You use this pattern a lot in the tests, but I'm not too familiar with it. Could you explain what the tests are checking with `assert_called_once()` and `assert_called_once_with`. 🤔

Reply:
I should be using `assert_called_once_with` here as well, but this test is simplified for now. In this and other tests, I'm ensuring that a method is called, but I don't necessarily need that method to be called or test any further than that, as other tests would cover the rest. In this case, `get` calls `client.get` and `manager.handle_response`, and all I need to ensure is that they are being called with the correct parameters. Thoughts?

Reply:
makes sense. I suppose richer functionality will be tested in future tests